enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,502 @@
+used to configure `memory.swap.max` to zero on the system slice parent cgroup.
+The configuration is done via systemd for the sake of consistency.
+
+* Preventing SWAP traffic I/O saturation
+
+One risk of heavy swapping is to saturate the disk bus with swap traffic,
+potentially preventing other processes from performing I/O.
+
+In order to ensure that system services are able to perform I/O, the
+machine config is configuring `io.latency=50` for the `system.slice` in order
+to ensure that its I/O requests are prioritized over any other slice.
+This is, because by default, no other slice is configured to have
+`io.latency` set.
+
+* Node memory pressure handling
+
+Dealing with memory pressure on a node is differentiating the TP fom GA.
+
+** **Technology Preview** - `memory.high` is set on the `kubepods.slice`
+  in order to force the node to swap, once the `kubepods.slice` memory is
+  filling up. Only once swap is full, the system will cross `memory.high`
+  and trigger soft evictions.
+
+  * Pro
+    * Simple to achieve.
+  * Con
+    * A lot of memory pressure has to be present in order to trigger
+      soft eviction.
+    * Once `memory.high` is reached, the whole `kubepods.slice` is throttled
+      and cannot allocate memory, which might lead to applications crashing.
+
+** **General Availability** - Memory-based soft eviction is going to
+  be disabled, in favor of enabling swap-based hard evictions, based on new
+  swap traffic and swap utilization eviction metrics.
+
+  * Pro
+    * Eviction on the basis of swap pressure, not only memory pressure.
+    * [LLN] applies, because all pods share the nodes memory
+  * Con
+    * Swap-based evictions are made through a 3rd party container, which means
+      it has to be done through an API-initiated eviction.
+
+###### Hypervisor/OS level assistance 
+
+Swap allows us to increase the
+virtual address space. This is required in order to reduce the
+likelihood for workloads to OOM significantly.
+
+Besides _increasing_ the virtual address space, there are two
+mechanisms which help to reduce the memory footprint of a workload.
+
+They are not being discussed in depth here, but are mentioned, as they
+help to reduce or avoid memory pressure, which in turn leads less usage
+of swap.
+
+The mechanisms are:
+* _Free Page Reporting ([FPR])_ - A mechanism to free memory on the
+  hypervisor side if it is not used by the guest anymore.
+* _Kernel Samepage Merging ([KSM])_ - A mechanism to merge/deduplicate
+  memory pages with the same content. This is particularly helpful
+  in situations where similar guests are running on a hypervisor.
+  It has however some security implications.
+
+##### VM higher-density: Differences between Technology Preview vs GA
+
+|                              | TP                  | GA                  |
+|------------------------------|---------------------|---------------------|
+| SWAP Provisioning            | MachineConfig       | MachineConfig       |
```

Comment:
GA: operator?

Reply:
In CNV higher density with wasp is already GA.

Reply:
One clear advantage of having the swap operator contributed directly to Openshift rather than CNV->Openshift is  less API transitions for the end user.

Reply:
@haircommander @fabiand I've been thinking about it again, and actually no API change should happen. Its just a matter of deployment. In CNV we will deploy it with HCO, later on it will be deployed by OLM.